[Gwee Yeu Chai] iP

gweeyc opened the pull request:
# LisGenie Chatbot Program (iP)
> "The way to get started is to quit talking and begin doing". - [Walt Disney](https://www.forbes.com/sites/lewishowes/2012/07/17/20-business-quotes-and-lessons-from-walt-disney/?sh=39926494ba9a)
## A Fun CLI-Style App
**LisGenie Chatbot** frees your mind of having to remember things you need to do. It's,
* text-based
* easy to learn
* ~~FAST~~ *SUPER* FAST to use

All you need to do is,

1. clone its repo from [here](https://github.com/gweeyc/ip.git).
1. open the project in your favorite Java IDE app.
1. run the program.
1. add your tasks using CLI commands, e.g.
   1. todo `<task description>`
   1. deadline `<task description>< /by ><datetime>`
   1. event `<task description>< /at ><datetime>`
1. let it manage your tasks for you 😀 

And it is **FREE**!

Features:

- [x] Managing tasks
- [ ] Managing deadlines
- [ ] Reminders (coming soon)
* * *
If you Java programmer, you can use it to practice Java also. Here's the main method:

```Java
public static void main(String[] args) {
        // Greeting screen display
        String logo = " ____        _        \n"
                + "|  _ \\ _   _| | _____ \n"
                + "| | | | | | | |/ / _ \\\n"
                + "| |_| | |_| |   <  __/\n"
                + "|____/ \\__,_|_|\\_\\___|\n";
        System.out.println("Hello from\n" + logo);
        // Greet user
        greet();
        // Get user input
            Scanner in = new Scanner(System.in);
            String input;
            boolean isBye = false;
            // conversions loop
            do {
                System.out.printf("%nMasterOm : ");
                input = in.nextLine().trim();
                drawALine();

                if (input.isEmpty()) {
                    // flush buffer
                    in.nextLine();
                    echoEmptyInput();
                }

                try {
                    isBye = echoBye(input);
                } catch (DukeException e) {
                    System.out.println(e.getMessage());
                }finally {
                    drawALine();
                }
            } while (!isBye);

            in.close();
    }
```

Commit:
Add Gradle support

gweeyc committed:
Duke Level-0: Greet (skeletal version)

gweeyc committed:
Duke Level-1: Greet, Echo, Exit (skeletal version)

gweeyc committed:
Duke Level-2: Add, List

gweeyc committed:
Duke Level-3: Mark as Done

gweeyc committed:
Duke Leve-4: ToDo, Event, Deadline

Replace Scanner with better String split() method under Conversation runs section

Create a package ip.duke for Duke suite of files

Conform code to standards

gweeyc committed:
Duke A-TextUiTesting: Automated Text UI Testing (optional)

Run runtest.bat confirms actual/expected results agree (ignore line artifacts).

Minor change: in dialogue format to display names of MasterOm and LisGenie (Duke.java).

gweeyc committed:
A-CodingStandard: Follow the Coding Standard

Tweak all code to comply with the stipulated coding standard.

gweeyc committed:
Duke Level-5: Handle Errors

Added DukeException.java to contain custom exception class for Duke App

Adjust echoBye method to throw DukeException, try out exception for todo task

Minor edit to main method to catch DukeException & minimize number of try blocks

gweeyc committed:
Duke Level-5: Handle Errors All Done

Added Throwable parameter to DukeException class to store root error type

Done the main task commands error handling via custom Duke Exceptions

Corresponding edits to affected echo messages

Minor edit of horizontal line draws placements

gweeyc committed:
Merge branch 'branch-Level-5'

gweeyc committed:
Duke A-CodeQuality: Improve Code Quality (01)

Rename variables names to conform to stipulated standard

Space out chucks of related code for improved readability

gweeyc committed:
Improve Code Quality (02)

Use spaces to structure the code logically

Do KISS & SLAP as much as possible

gweeyc committed:
Merge branch 'branch-A-CodeQuality'

gweeyc committed:
Duke A-AbstractClasses: Use Abstract Classes

Change Task class to an abstract class

Change setId() method to abstract

Edit subclasses to implement setId()

gweeyc committed:
Duke A-AbstractClasses: Done

Add another abstract method getDescription() to Task class

Implement the method in Todo class

gweeyc committed:
Merge branch 'branch-A-AbstractClasses'

gweeyc committed:
Duke A-Packages: Organize into Packages

Move DukeException class into new package ip.duke.exceptions

Import the new package into Duke.java (to use DukeException)

gweeyc committed:
Organize into Packages: Done

Added Task class into ip.duke.task package

Add import codes for use in Duke and Todo classes

gweeyc committed:
Merge branch 'branch-A-Packages'

lowyanshuang reviewed (commented):
### Short and Sweet
Dont find any coding standard you have violated. And I like how you separate into different functions so that it is easier to reuse the code. Overall is short and easy to understand. Good work!

lowyanshuang reviewed (commented):
### Some Personal Additional Questions
Just wondering since the task is limited to 100, is it possible to not set the size of the task array and expand as you go?
Also for `case ""` i feel that this rather redundant, why not include it under default?

gweeyc commented:
Thanks yanshuang for the comments. Yes tasks size I can set to a constant in my Duke.java file instead, so the array size be not fixed at 100 only. But since the iP specification says max size = 100, so I just set it accordingly!

As for the case "", I was thinking of giving precise feedback for the different possible user inputs that can come thru'...so my chatbot replies to the user will alert user as to what he/she had entered wrongly, so can take corrective action in the next try. But I like the suggestion, offering a possible different view!

thisisdax commented on src/main/java/ip/duke/Duke.java:
```diff
@@ -0,0 +1,227 @@
+        }
+    }
+
+    private static boolean echoBye(String input) {
```
You can consider renaming this method to have more clarity.

Perhaps "handleInput" might be a better naming convention. 👍

thisisdax commented on src/main/java/ip/duke/Duke.java:
```diff
@@ -47,51 +47,54 @@ public static void main(String[] args) {
-                input = words[0].trim();
+            if(command.indexOf(" ") > 0){
+                words = command.split(" ", 2);
+                command = words[0].trim();
```
Since we are using split(" ") string method to split up the words delimited by spaces, it will naturally remove spaces from each of the elements returned.

```
String wordOne = "Hello World"
String[] splitWords = wordOne.split(" ")
// splitWords[0] == "Hello" and splitWords[1] == "World"

// In the case of multiple spaces (e.g. double spaces)
String wordTwo = "Hello  World"
splitWords = wordTwo.split(" ")
// splitWords[0] == "Hello"
// splitWords[1] == " "
// splitWords[2] == "World"
```

The trim() method removes whitespace from both ends of a string and since split handled that for you, you can save some extra steps here.

thisisdax reviewed (commented):
Good stuff! Couldn't find much else that I would do otherwise.

gweeyc committed:
Refactor boolean method name "echoBye" to "hasTask" - for better code standards

Simplify echoNoBy and echoNoAt method further

Remove redundant trim() method from words[0]

gweeyc committed:
A-Collections

Use LinkedHashSet instead of array to store tasks
- preserver input order, get constant time operations (non-iterative)

gweeyc committed:
Refactor code further

-remove duplicate codes.
-implement SLAP for methods, switch-cases.

gweeyc committed:
Merge pull request #1 from gweeyc/branch-Level-6

Branch level 6 new merge with master

gweeyc committed:
Refactor Task and subclass dependency construction

-better efficiency and clearer OOP application
-Change id String type to char
-created some write to file prep code in various class files
-added a blank tasks.txt file

gweeyc committed:
Edited Duke and Todo files for file writes

-corrected Todo's toFileString() format for better OOP usage.

gweeyc committed:
Rename remove method to drop to avoid Java API confusion

Created getTasksFromFile & createTask methods
- to load data to RAM from saved data tasks.txt file on start.

Simple test run / creation of tasks.txt successful.

gweeyc committed:
Change code from LinkedHashSet to LinkedHashMap usage
- get constant time for delete and update methods

Add try-catch code for switch block in createTask method

gweeyc committed:
JavaDoc comments done for Duke.java key methods

gweeyc committed:
Duke A-JavaDoc : JavaDoc

Created JavaDoc comments for various methods & classes

gweeyc committed:
Merge branch 'branch-A-JavaDoc'

gweeyc committed:
Merge branch 'master' into branch-Level-7

Conflict resolved

*A bug in listing function of LinkedHashMap to fix in next commit

gweeyc committed:
Debug successful: listing and write to file data agree

Replaced LinkedHashMap with LinkedHashSet yields
easier track of dynamic input sequence positioning on database changes

Created getItem method to extract Task item via its indexed sequence order

Replace buggy INDEX with TASKS.size() - 1

gweeyc committed:
Merge branch 'branch-Level-7'

gweeyc committed:
Duke A-Jar: Create a JAR File

ip JAR FIle created

gweeyc committed:
Duke A-MoreOOP : Use More OOP

gweeyc committed:
Leve-8 : add find and list find methods

gweeyc committed:
added javadoc for Ui class

gweeyc committed:
A-JUnit: Add JUnit Tests

Done 3 tests for Todo class.

Remove the Wrapper.java as not needed.

gweeyc committed:
Added 3 JUnit 5 tests for Deadline class

gweeyc committed:
Merge pull request #3 from gweeyc/branch-Level-9

Branch level 9 : add in a find method for user

gweeyc committed:
Merge branch 'master' into branch-A-JUnit

gweeyc committed:
Merge pull request #4 from gweeyc/branch-A-JUnit

Branch a jUnit 5 to merge

gweeyc committed:
Remove unwanted backup.txt.

Call hasTask method as static method of Parser Utility Class.

Added JavaDoc for Utility classes.

gweeyc committed:
Duke Level-8: Dates and Times

Configured Dateline Class Constructor to use LocalDate parse method.
Event Class inherits ability.

gweeyc committed:
Update README.md

Added a section spelling out the supported CLI commands to use

gweeyc committed:
Set theme jekyll-theme-time-machine

gweeyc committed:
Set theme jekyll-theme-cayman

gweeyc committed:
Set theme jekyll-theme-cayman

gweeyc committed:
Set theme jekyll-theme-cayman

gweeyc committed:
Set theme jekyll-theme-cayman

gweeyc committed:
Edited README.md with inline code format for CLI commands

gweeyc committed:
Merge branch 'master' of https://github.com/gweeyc/ip

gweeyc committed:
Update README.md

remove Java code formatting to make compatible with webpage display

gweeyc committed:
Set theme jekyll-theme-dinky

gweeyc committed:
Set theme jekyll-theme-architect

gweeyc committed:
Update README.md

gweeyc committed:
Duke A-Release: Release

Created main.jar that takes care of data folder creation and database save files.

gweeyc committed:
Merge branch 'master' of https://github.com/gweeyc/ip

gweeyc committed:
Fix a bug for single digit inputs (without a '0' in front) parsing
- in LocalDate.parse method of Deadline.java file,
- also allow another date input format of "yyyy/M/d"

gweeyc committed:
Update README.md

Added comments to help guide user for deadline and event input strings

gweeyc committed:
Update README.md

Corrected type on last line #NB: -

gweeyc committed:
Update README.md

Add details to the command descriptions

gweeyc committed:
tested more Deadline and Event tasks input

gweeyc committed:
Change the delimiter from ":" to # to avoid clash with time format also using ":"

Edited affected files to work with change.

Modified createTask mehtod to throw DukeException in TaskList.java.

Modified  showloadingError method in Ui.java to show more related error messages

gweeyc committed:
Edited comments to conform to stipulated standards

gweeyc committed:
Set Duke Class file and backup file paths as class constant.

gweeyc committed:
Update README.md

Edited grammar and added info on warning on database full.

gweeyc committed:
Obviate trim() loop from createTask method via new split delimiter " # ".

gweeyc committed:
Refine Ui's echoFind method to use item.toString() over item.getDescription() to widen search capability.

gweeyc committed:
redo file paths assignment for better backup route.

gweeyc committed:
Merge remote-tracking branch 'origin/add-gradle-support'

* origin/add-gradle-support:
  Add Gradle support

gweeyc committed:
Check coding style violations via gradle checkstyleTest & checkstyleMain.

gweeyc committed:
Fixed code style to conform to checkstyle-Main & -Test.

Altered the checkstyle.xml to omit exhausive prompt for Javadoc for every method.

gweeyc committed:
Corrected README.md tasks' classifications.

Added JavaFx dependencies in build.gradle.

gweeyc committed:
Successfully added the JavaFX GUI to Duke App.

gweeyc committed:
Corrected build.gradle and Manifest.mf code lines to point to Launcher Class.

gweeyc committed:
Set up CI, tested checkstyle working ok locally.

Edited some newlines in LisGenie's tasks replies.

gweeyc committed:
testing CI setup 1.

gweeyc committed:
Done C-FriendlierSyntax extension for all Duke CLI commands.

gweeyc committed:
Added Exception handling to LisGenie App.

Corrected path to Launcher.java.

indent long code lines display & add parameters null checks for Image method for MainWindow.java.

Added parameters null checks for Ui methods.

gweeyc committed:
Added System.setOut in main method to point to System.out for Duke.java file alone:
- to run in CLI mode if needed.

gweeyc committed:
Created a visually improved textbox fromat for LisGenie Chatbot.

gweeyc committed:
Add Ui.png

Added the LisGenie GUI chatbot dialogue box screenshot.

gweeyc committed:
Update and rename README.md to UserGuide.md

Edited for the added command short form availability.

gweeyc committed:
First Draft of Developer Guide

pending finalization with a second edit.

gweeyc committed:
Added App Class UML Diagram file

gweeyc committed:
Added a Sample ChatBox Picture

gweeyc committed:
Update DeveloperGuide.md

Completed the DG doc.

gweeyc committed:
Create README.md

gweeyc committed:
Add files via upload

gweeyc committed:
Update DeveloperGuide.md

Copied the graphic files into doc folder.

gweeyc committed:
Delete LisGenie_UML.png

gweeyc committed:
Delete lisgenieChatBox.png

gweeyc committed:
Update DeveloperGuide.md

Corrected spelling mistakes.

gweeyc committed:
Update DeveloperGuide.md

Added Non-functional Req numbered line 6.

gweeyc committed:
Update _config.yml

added title

gweeyc committed:
Update _config.yml

gweeyc committed:
Update README.md

gweeyc committed:
Update DeveloperGuide.md

Improved grammar and sentences' expressions.
